core-bundle/tests/Contao/ModelTest.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Tests\Contao;
+
+use Contao\CoreBundle\Tests\TestCase;
+use Contao\System;
+use Symfony\Component\Filesystem\Filesystem;
+
+/**
+ * @group legacy
+ */
+class ModelTest extends TestCase
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $container = $this->getContainerWithContaoConfiguration();
+        $container->setParameter('contao.resources_paths', $this->getTempDir());
+
+        (new Filesystem())->mkdir($this->getTempDir().'/languages/en');
+
+        System::setContainer($container);
+    }
+
+    public function testThrowsExceptionForNonExistantRelations(): void
+    {
+        $this->expectException(\Exception::class);
+
+        // ContentModel could be any other Model class extending from \Contao\Model
+        $model = new \Contao\ContentModel();
+        $model->getRelated('foo');
+    }
+
+    /**
+     * @dataProvider brokenRelationsProvider
+     */
+    public function testIncludesTableAndRelationNameForNonExistantRelationsInException(string $table, string $relation): void
+    {
+        // We expect to find the table name and the relation in the exception,
+        // e.g. tl_foo.bar for the relation "bar" of table "tl_foo".
+        $expectedExceptionContent = $table.'.'.$relation;
+        $this->expectException(\Exception::class);
+        $this->expectExceptionMessageMatches('/'.preg_quote($expectedExceptionContent).'/');
+
+        // Sadly, we cannot use a mock of \Contao\Model for abstract testing
+        // since PHPUnit mocks cannot call or mock static methods (in this case,
+        // static::getTable() within the Model), so use a real model
+        // class and dynamically set the table name via reflection.
+
+        // ContentModel could be any other Model class extending from \Contao\Model
+        $mock = new \Contao\ContentModel();
```

Comment:
Instead of using reflection here you can use the `Adapter` class: https://github.com/contao/contao/blob/4.9/core-bundle/src/Framework/Adapter.php

Reply:
I don't see how this would be possible with the Adapter class since the problematic call to `static::getTable()` is right within the class instance.

While the `Adapter` would help for an external call via `$adapter->getTable()`, we need to mock the static method within the subject itself – the Model instance we call `getRelation` on, which then  calls `static::getTable()` internally. 

Sadly, mocking a static method is not possible with PHPUnit mocks (anymore), so I think reflection is the easiest way here.

On the other hand, we could use the `Adapter` within `Model::getRelated()` to resolve the table name in a testable way, but this doesn't feel like the right way either looking at the amount of static calls within the `Model` class to itself (`static::getTable`, `static::getClassFromTable` etc.).

Am I missing something?

Reply:
Hmm I see. I don't think we ever tested legacy code because of this... Let's see what the others think about.

Reply:
I don't think we need a test for a modified exception message. Just delete the `ModelTest` class again.

Reply:
Sounds reasonable. I have updated the pull request and removed the unit-test.